### ISSUE TITLE
Data Conversion Problem TPVFormer -> OCC-VO

### ISSUE THREAD

**Issue by whoamibro** [USTCLH/OCC-VO#7]
Hello, thank you very much for releasing the code for the visual odometry method based on occupancy.

However, the occupancy produced by TPVFormer seems quite different in format from the data in samples/occupancy. 
After reviewing your released code to understand the input data structure, 
I converted the occupancy estimated by TPVFormer into the input format required by OCC-VO, and removed all points corresponding to the empty class(index 17 generated from the TPVFormer). 
But after doing so, the number of input points I get is only between 1,000–3,000, instead of 10,000–20,000 as in the samples folder.

When I run the OCC-VO algorithm with this data, the visual odometry estimation doesn’t seem to work properly. If the samples stored in samples/occupancy/scene-0101 and the experiment results were obtained using occupancy data from TPVFormer, could you please share the code you used to convert TPVFormer’s output (with shape 1 x 18 x 100 x 100 x 8) into the N x 4 input format?

I used the model file tpv04_occupancy_v2.pth provided by the authors when estimating with TPVFormer and nuScenes test dataset.

**Comment by USTCLH** [comment]
Hello, thanks for your attention.

The issue you mentioned mainly comes from differences in the datasets. In OCC-VO,  the TPV-Former we used is retrained on Occ3D-nuScenes, rather than tpv04_occupancy. The latter has a significantly smaller number of voxels compared to the former, which is the root cause of the problem. For instructions on how to retrain on Occ3D, please refer to [the official Occ3D repository](https://github.com/[Tsinghua-MARS-Lab/Occ3D](https://github.com/Tsinghua-MARS-Lab/Occ3D)) or other works on Occ3D dataset.

As for converting the Occupancy output from .npz files to our .npy format, the code is actually quite straightforward, as shown below:
```
import os
import copy
import numpy as np
from tqdm import tqdm
from nuscenes.nuscenes import NuScenes

class gridtool():
    def __init__(self, grid_resolution, grid_size, grid_center):
        self.resolution = grid_resolution
        self.size = grid_size
        self.center = grid_center

        # Map Occupancy to points
        x_range = np.arange(self.grid_size[0]) - self.grid_center[0] + 0.5
        y_range = np.arange(self.grid_size[1]) - self.grid_center[1] + 0.5
        z_range = np.arange(self.grid_size[2]) - self.grid_center[2] + 0.5
        x, y, z = np.meshgrid(x_range, y_range, z_range, indexing='ij')
        x = x * self.grid_resolution
        y = y * self.grid_resolution
        z = z * self.grid_resolution

        self.grid_pos = np.stack((x, y, z, np.zeros_like(x)), axis=-1)

    def convert_pcd(self, input, mask):
        tmp = copy.deepcopy(self.grid_pos)
        tmp[mask, 3] = input[mask] 

        return tmp[mask]

if __name__ == '__main__':   
    path_nusc = 'your_path_to_nuscenes/trainval'
    # path_nusc = 'your_path_to_nuscenes/test'
    path_src = 'your_path_to_results/tpv_results'
    path_dst = 'your_path_to_outputs/tpv'
    path_dst2 = 'your_path_to_outputs/tpv_gt'

    grid_resolution = 0.4
    grid_size = (200, 200, 16)
    grid_center = (100, 100, 2.5)
    grid = gridtool(grid_resolution, grid_size, grid_center)

    nusc = NuScenes(version='v1.0-trainval', dataroot=path_nusc, verbose=False)

    file_list = os.listdir(path_src)
    file_list.sort() 

    last_scene = ''
    count = 0
    for file in tqdm(file_list):
        result = np.load(path_src + '/' + file)

        pred = result['pred']
        gt = result['gt_semantics']
        mask = (result['mask_camera'] != 0)

        mask_pred = copy.deepcopy(mask)
        mask_pred[pred >= 17] = False
        pcd_pred = grid.convert_pcd(pred, mask_pred) 
        # pcd_pred = np.asarray(pcd_pred[:, 0:3], np.float32)

        mask_gt = copy.deepcopy(mask)
        mask_gt[gt >= 17] = False
        pcd_gt = grid.convert_pcd(gt, mask_gt)
        # pcd_gt = np.asarray(pcd_gt[:, 0:3], np.float32)

        sample = nusc.get('sample', str(result['sample_token']))
        scene = nusc.get('scene', sample['scene_token'])

        os.makedirs(path_dst + '/' + scene['name'], exist_ok=True)
        os.makedirs(path_dst2 + '/' + scene['name'], exist_ok=True)

        if scene['name'] != last_scene:
            count = 0

        last_scene = scene['name']

        np.save(path_dst + '/' + scene['name'] + '/' + str(count) + '.npy', pcd_pred)
        np.save(path_dst2 + '/' + scene['name'] + '/' + str(count) + '.npy', pcd_gt)

        count = count + 1
    
```

**Comment by USTCLH** [comment]
In addition, all the parameters of OCC-VO are tuned based on the occupancy results obtained on Occ3D. OCC-VO can be used for other datasets, but parameter adjustments may be required. However, for these datasets, we expect the occupancy resolution to be more fine-grained in order to achieve good performance. In this respect, the resolution of tpv04_occupancy is too coarse, and I think it may be difficult to obtain accurate results with it.

**Comment by whoamibro** [comment]
Thank you for kind and quick reply. 
Then, could you share the configuration you used when retraining TPVFormer on the occ3d nuScenes dataset, and the OCC-VO configuration and hyper-parameter values that you adjusted for that dataset?

**Comment by USTCLH** [comment]
In fact, for the configuration, you only need to adjust the occupancy settings (such as range and resolution) for the Occ3D dataset. We do not make any structural modifications to TPVFormer. As for the corresponding OCC-VO settings and hyper-parameters, the ones provided in our repository are exactly what we used. Moreover, if your goal is simply to reproduce the performance of OCC-VO on Occ3D-nuScenes, the [Google Drive link](https://drive.google.com/drive/folders/18PW9KanUN5mwKxRWn3XcFitpvRCXg4Lk) in README already contains the occupancy data on the validation set that OCC-VO requires as input.

**Comment by whoamibro** [comment]
Okay I'll try as you say! I really appreciate your answers.
